compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/DistributeMmaToLanes.cpp

```diff
@@ -0,0 +1,150 @@
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.h"
+#include "iree/compiler/Codegen/Dialect/GPU/Transforms/Passes.h"
+#include "iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/TileUsingInterface.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Interfaces/LoopLikeInterface.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler::IREE::GPU {
+
+#define GEN_PASS_DEF_DISTRIBUTEMMATOLANESPASS
+#include "iree/compiler/Codegen/Dialect/GPU/Transforms/Passes.h.inc"
+
+namespace {
+struct DistributeMmaToLanesPass final
+    : impl::DistributeMmaToLanesPassBase<DistributeMmaToLanesPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+struct ConvertToMultiMma final : OpInterfaceRewritePattern<linalg::LinalgOp> {
+  using OpInterfaceRewritePattern::OpInterfaceRewritePattern;
+  LogicalResult matchAndRewrite(linalg::LinalgOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    auto loweringConfig =
+        getLoweringConfig<IREE::GPU::LoweringConfigAttr>(linalgOp);
+    if (!loweringConfig) {
+      return failure();
+    }
+    IREE::GPU::MmaInterfaceAttr kind = loweringConfig.getMmaKind();
+    if (!kind) {
+      return failure();
+    }
+    if (failed(convertContractionToMultiMma(rewriter, linalgOp, kind))) {
+      return failure();
+    }
+    return success();
+  }
+};
+
+LogicalResult fuseProducersGreedily(RewriterBase &rewriter,
+                                    scf::ForallOp laneForall) {
+
+  std::deque<tensor::ExtractSliceOp> candidates;
+  laneForall->walk([&](tensor::ExtractSliceOp extractSliceOp) {
+    auto producer = extractSliceOp.getSource().getDefiningOp<TilingInterface>();
+    if (producer && producer->getBlock() != laneForall.getBody()) {
+      candidates.push_back(extractSliceOp);
+    }
+  });
+
+  SmallVector<LoopLikeOpInterface> loops = {laneForall};
+
+  OpBuilder::InsertionGuard g(rewriter);
+  while (!candidates.empty()) {
+    // Traverse the slices in BFS fashion.
+    tensor::ExtractSliceOp candidateSliceOp = candidates.front();
+    candidates.pop_front();
+
+    // Materialize the slice of the producer in place.
+    std::optional<scf::SCFFuseProducerOfSliceResult> fusedProducer =
+        scf::tileAndFuseProducerOfSlice(rewriter, candidateSliceOp, loops);
+    if (!fusedProducer)
+      continue;
+
+    // We have no way to know whether a multi-use value can be yielded from the
```

Comment:
You can do it through dominators. You can look at the uses of the producer. any use that is dominated by the loop could be replaced with value yielded from the loop. Its OK to leave that out for now though.

Reply:
That wasn't the reason I left this comment. The problem is that the tile of the producer might be recomputed multiple times across threads (because this is fusing into the operand of a matmul) so we won't know how to construct the result (a la tensor.parallel_insert_slice) properly.